### DISCUSSION TITLE
Generate FMU in DAE Mode

### DISCUSSION THREAD

**Discussion by b-mare** [OpenModelica/OpenModelica#11807] [Q&A]
Hi,

Would it be possible to generate an FMU containing a DAE instead of an ODE? I saw the simulation flag -daeMode, but did not find how I can pass this to the FMU generation. 
Likewise I would like to know if/how I can create an FMU using sparse linear solvers like klu to be able to exploit the sparsity of the model in the FMU. 
Is there anyone who has experience with this?

Kind regards
bmare

**Comment by casella** [comment]
FMUs are inherently based on an internal state-space representation, in particular if you want to use FMI/ME. As far as I understand, they really do not fit the daeMode framework.

Using sparse solvers to solve implicit equations when computing derivatives is a different story. Currently we automatically switch to sparse solvers if the implicit systems are large and sparse enough. I'm not sure we also do it for FMUs. @AnHeuermann could you please comment on that?

**Reply by AnHeuermann** [reply]
I don't think the sparse solvers are added to the FMU to keep it small and simple to compile.

**Reply by casella** [reply]
We could add them behind some flag, please let us know if you have a specific need for that

**Comment by jgillis** [comment]
Fwiw, this paper discusses both strategies with OpenModelica: Braun, W., Casella, F., & Bachmann, B. (2017). [Solving large-scale Modelica models: new approaches and experimental results using OpenModelica](https://re.public.polimi.it/bitstream/11311/1065357/1/2017-BraunCasellaBachmann.pdf). In LINKÖPING ELECTRONIC CONFERENCE PROCEEDINGS (PRINT) (pp. 557-563). Linkoping University Electronic Press.

**Reply by casella** [reply]
@jgillis I'm aware of that paper 😄 

The point is: OMC can generate models where the DAE is kept together as such, and passed on to a DAE solver such as IDA; however, the FMI standard doesn't have this provision, because the underlying mathematical structure of the system is explicit state space. If you generate an FMI-CS, this will need to compute the derivatives of the states given inputs, states and time, which is not what a DAE formulation does.

In fact, we could use daeMode for FMI-CS, using daeMode and IDA to advance the simulation by one step. But that would then only work in a fixed-time-step co-simulation framework. And, it would require non-trivial implementation work, since the solvers that are currently supported inside FMI-CS are ODE solvers (forward Euler and CVODE), not DAE solvers.

Would that be interesting for you?

**Reply by jgillis** [reply]
Woopsie, I bet you are 😄

Our interest (I'm working together with @b-mare ) would be in ME only, not CS.
I probably got confused wrongfully assuming the paper was about ME.

On the topic of DAEs, how about letting OMC treat algebraic constraints as outputs and algebraic variables as inputs (perhaps with some convention on labeling such that the FMU/ME tool can treat these specially)? Is that something that could be achieved with minor modifications in OMC? Do you see fundamental issues which such a route?

We can simplify the situation by assuming we have no discrete dynamics..

**Reply by casella** [reply]
The paper was about the regular C runtime, not FMI.

There have been talks about extending FMI to support DAEs, but I'm not aware about any official development. Until then, FMI tools will not be able to deal with DAEs. Of course we could tweak the way we generate FMIs, but if FMI tools can't handle them, what's the point?

**Reply by casella** [reply]
The question is, what are your actual requirements? Why would you like to have daeMode FMUs?

**Reply by jgillis** [reply]
Thanks for clarifying this point about C runtime. An option for us could be to work with that runtime instead of via FMU.
As for context, it's large scale optimal control with a sparse DAE implemented in Modelica.

**Reply by casella** [reply]
If you write down a clear mathematical specification of what would you need to get from the generated code, it may already be available. RTE is generating DAE code of power system model components using OpenModelica and then integrating them into their [Dynawo](https://dynawo.github.io/) open-source code, maybe you could do something similar

**Reply by b-mare** [reply]
How can we create this generated C code? Is the *buildModel* script to way to go or are there other options?

I looked into the Dynawo code. It seems instantiateModel() and translateModel() are sufficient. Is there a difference between doing instantiateModel + translateModel and buildModel?

**Reply by casella** [reply]
It should be the same. For some reason translateModel is not documented in the User's Guide (it probably should, maybe you can open an issue so we remember to fix that).

**Reply by b-mare** [reply]
We will try this approach.
If we encounter issues, I will start a new discussion.

Thanks.

**Comment by chrbertsch** [comment]
@jgillis There are proposals  for DAE support in FMI, but they did not make it into FMI 3.0, because they were not mature enough (there were e.g. open questions regarding initialization)
However, DAE support could be added in FMI 3.x. Currently there is no active working group on DAE support in FMI, but it could be revived.
You are welcome to describe your requirements in an issue on https://github.com/modelica/fmi-standard/issues

**Reply by casella** [reply]
Thank you @chrbertsch!
